### PR TITLE
Feature/33 custom field

### PR DESCRIPTION
**Pull request by timohuber**

### CODE REVIEW

File: pool/web/handler/admin_custom_fields.ml [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+      let group, id = CCString.take_drop (CCString.length field) key in
+      match CCString.equal field group with
+      | false -> None
+      | true ->
+        let key =
+          let open CCOption in
+          id
+          |> CCString.chop_prefix ~pre:"["
+          >>= CCString.chop_suffix ~suf:"]"
+          >>= fun l -> l |> encoder
```
Comment (mabiede) [inline pool/web/handler/admin_custom_fields.ml:23]: - `>>= encoder`
- if "value" is defined before this "match with" statement, could it directly be included?

File: pool/web/handler/admin_custom_fields.ml [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+let form ?id req =
+  let open Utils.Lwt_result.Infix in
+  let open Lwt_result.Syntax in
+  let result ({ Pool_context.tenant_db; _ } as context) =
+    Lwt_result.map_error (fun err -> err, "/admin/custom-fields")
+    @@
+    let flash_fetcher key = Sihl.Web.Flash.find key req in
+    let* custom_field =
+      match id with
```
Comment (mabiede) [inline pool/web/handler/admin_custom_fields.ml:56]: This seems to be something like `>|> CCOption.map_or ~default:(Lwt.return_none) Custom_field.find`

File: pool/web/view/page/page_admin_custom_fields.ml [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+      ?orientation
+      ?help
+      field
+      ~value:(custom_field |> CCOption.map_or ~default fnc)
+      ~flash_fetcher
+  in
+  let value = CCFun.flip (CCOption.map_or ~default:"") custom_field in
+  let input_by_lang ?(required = false) field value_fnc =
+    let open Pool_common in
+    let group_class = Component.Elements.group_class [] `Horizontal in
```
Comment (mabiede) [inline pool/web/view/page/page_admin_custom_fields.ml:31]: `Component` is already opened

File: pool/web/view/page/page_admin_custom_fields.ml [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+                  (if required
+                  then Format.asprintf "%s *" label_text
+                  else label_text)
+              ]
+          ; input_element
+          ])
+      sys_languages
+  in
+  let name_inputs =
+    input_by_lang ~required:true Pool_common.Message.Field.Name (fun lang ->
```
Comment (mabiede) [inline pool/web/view/page/page_admin_custom_fields.ml:77]: `Message` is defined as `Pool_common.Message` at the top

File: pool/web/view/page/page_admin_custom_fields.ml [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+  let checkbox_element ?orientation ?help ?(default = false) field fnc =
+    checkbox_element
+      language
+      ?orientation
+      ?help
+      field
+      ~value:(custom_field |> CCOption.map_or ~default fnc)
+      ~flash_fetcher
+  in
+  let value = CCFun.flip (CCOption.map_or ~default:"") custom_field in
```
Comment (mabiede) [inline pool/web/view/page/page_admin_custom_fields.ml:28]: Bad "function" name, there are fields called "value", very hard to spot where the function is called. I found only one spot, may include it direclty.

**Review by mabiede — CHANGES_REQUESTED** [review]

File: pool/app/custom_field/repo/repo_entity.ml [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+end
+
+module Name = struct
+  include Name
+
+  let t =
+    let encode t =
+      t |> yojson_of_t |> Yojson.Safe.to_string |> CCResult.return
+    in
+    let decode t =
```
[the remark this answers is not in the document]
Comment (timohuber) [reply pool/app/custom_field/repo/repo_entity.ml:33]: Yes, that was the plan. But i obviously forgot to replace the functions.

**Pull request merged**